DPL: fix routing issues in forwarding

ktf opened the pull request:
If one (header, payload, ...) tuple in a MessageSet was to be copied,
all the subsequent ones would have been copied.

If one (header, payload, ...) tuple got redirected to more than one destination,
all the subsequent ones would have been redirected there.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/AliceO2Group/AliceO2/pull/14859).
* #14910 (2 commits)
* __->__ #14859 (2 commits)

ktf commented:
@shahor02 @davidrohr @Barthelemy @knopers8 this might explain some of the issues you have seen in async reconstruction of pp data on the EPN. There was for sure some flawed logic, I am not sure if it could be triggered in normal circumstances apart from EoS or multi proxy setups when wildcards are involved. Unless you are 100% happy with this, I want to first write a unit test for the new forwarding logic to make sure all my assumptions are correct.

davidrohr commented on Framework/Core/src/DataProcessingDevice.cxx:
```diff
@@ -658,46 +588,68 @@ auto decongestionCallbackLate = [](AsyncTask& task, size_t aid) -> void {
+      // We need to find the forward route only for the first
+      // part of a split payload. All the others will use the same.
+      // but always check if we have a sequence of multiple payloads
+      cachedForwardingChoices.clear();
```
I don't understand, how does the caching work for multi-parts now?

ktf replied:
I renamed the variable to avoid confusion. I think the ability to cache the forwarding choice was either wrong or only correct in a past where wildcards were not there. To exemplify:

- Assume you have a wildcard InputSpec for `SRC/*`.
- This will match `SRC/A` and `SRC/B` and put them in the same messageSet.
- There is no guarantee that both `SRC/A` and `SRC/B` are to be sent to the same place.

The above only worked so far because I assume in general a wildcard is used for sinks and therefore it is last.

For multiparts, it is enough to iterate over the `messageSet.getNumberOfPayloads(pi)` as it's done later in the code.

ktf replied:
Ok, I see what you mean now. You are talking about the split payloads in the form:

```
[Header1][payload0][header1][payload1]... etc
```

indeed for those I need to have the cache to stay around in the new version... Will do it in a bit.

ktf replied:
Updated with the old logic and improved description.

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 57e7546821edf6be08de24fc0db2c3772947aae2 at 2025-11-29 22:54:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14859/57e7546821edf6be08de24fc0db2c3772947aae2/build_O2_fullCI_slc9/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for aa79411ae724556608142b01710f78fe121da4db at 2025-12-08 15:06:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14859/aa79411ae724556608142b01710f78fe121da4db/build_O2_fullCI_slc9/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 35d6915bb21367bcd10771f2f3e6d8808d952db6 at 2026-01-06 21:05:
```
## sw/BUILD/O2-latest/log
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:48:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:98:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:149:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:205:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:263:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:328:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:391:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:469:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:532:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:614:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:660:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/test/test_ForwardInputs.cxx:705:77: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/src/DataProcessingHelpers.cxx:224:6: error: no declaration matches 'std::vector<fair::mq::Parts> o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool, bool)'
/sw/SOURCES/O2/14859-slc9_x86-64/0/Framework/Core/src/DataProcessingDevice.cxx:595:70: error: no matching function for call to 'o2::framework::DataProcessingHelpers::routeForwardedMessages(o2::framework::FairMQDeviceProxy&, std::vector<o2::framework::MessageSet>&, bool&, bool&)'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14859/35d6915bb21367bcd10771f2f3e6d8808d952db6/build_O2_fullCI_slc9/pretty.html).

ktf committed:
DPL: fix routing issues in forwarding

If one (header, payload, ...) tuple in a MessageSet was to be copied,
all the subsequent ones would have been copied.

If one (header, payload, ...) tuple got redirected to more than one destination,
all the subsequent ones would have been redirected there.

ktf committed:
DPL: more refactoring of the forwarding code

Use a single helper function to improve readability.
